salt/syspaths.py

```diff
@@ -88,11 +90,42 @@
 CLOUD_DIR = os.path.join(INSTALL_DIR, "cloud")
 BOOTSTRAP = os.path.join(CLOUD_DIR, "deploy", "bootstrap-salt.sh")
 
+
+def _get_windows_root_dir():
+    # Try to get the root directory location from the registry
+    # This key will be created by the NullSoft installer
+    # If salt is currently installed in C:\salt and the user performs an
+    # upgrade, then this key will be set to C:\salt
+    root_dir = salt.utils.win_reg.read_value(
+        hive="HKLM", key="SOFTWARE\\Salt Project\\salt", vname="root_dir"
```

Comment:
SOFTWARE\\Salt Project\\salt  should be `key=SOFTWARE\\Salt Project\\salt-minion`  (or salt-agent) this then allows for master on windows as well.

Reply:
The Uninstall Reg `HKEY_LOCAL_MACHINE\Software\Microsoft\Windows\CurrentVersion\Uninstall`  also supports setting `InstallLocation`

Reply:
> SOFTWARE\\Salt Project\\salt  should be `key=SOFTWARE\\Salt Project\\salt-minion`  (or salt-agent) this then allows for master on windows as well.

good point, also true for the file layout

Reply:
> The Uninstall Reg `HKEY_LOCAL_MACHINE\Software\Microsoft\Windows\CurrentVersion\Uninstall`  also supports setting `InstallLocation`

Yes, but the SEP demands a second key: root-dir. I think keeping install-dir and root-dir together as now is the better choice than reusing that key and separating them. Also, the msi can store an option as third key: REMOVE_CONFIG

Reply:
I don't think it's uncommon for the master and minion to have their config in the same location. It's handled that way on Linux (`/etc/salt`).

Reply:
So how would the salt master installer work, if it also could be installed in any directory on Windows,  would it override the setting of root_dir    On link they share /etc/salt  and then they have with own files and folders which are not shared. `SOFTWARE\\Salt Project\\salt-minion` and `SOFTWARE\\Salt Project\\salt-master` would match more to how /etc/salt is structured. Consider `/etc/salt` being `SOFTWARE\\Salt Project`  and `SOFTWARE\\Salt Project\\salt-minion` being `minion` file and `minion.d` folder.    The root_dir can be set differently  both the `minion` and `master` file within `/etc/salt`

Reply:
You also need to have `conf_file`  e.g. %ProgramData%/salt-minion/pki and optionally `pki_dir` which would be %ProgramData%/salt-minion/pki

Salt Unix/Linux /etc structure is /etc/salt/pki/<master|minion> however on windows it might be better    %ProgramData%/salt-minion/pki and %ProgramData%/salt-master/pki ; %ProgramData%/salt-minion/conf and %ProgramData%/salt-master/conf   which would allow the master to run as a different user.

Consider using REG_EXPAND_SZ type which will expand the %ProgramData% on read (assume it does not lead to security issues)

Anyway keep in mind how a master installer would work.

Reply:
@damon-atkins There wouldn't be a separate installer for the master. We would just include the salt-master binary in an installer called "Salt". The registry setting is just a default location to use when looking for the config. In the salt-minion service registration, for example, we pass the `-c` option and specify a config.

However, the `root_dir` is not configurable with the "Install Anywhere" installer. Install anywhere only applies to the binaries... basically the `bin` directory and supporting scripts, dlls, and the ssm.exe. Everything else will be in `ProgramData` by default. If you wish to put that in a location other than `ProgramData`, you'll need to pass a `-c` option and specify those settings in the minion/master config. The only other valid value for `root_dir` would be `C:\salt` if installed over an existing installation where everything is in the `c:\salt` directory and the user elects NOT to move it.

Reply:
@twangboy on Unix/Linux the master does not need to be installed on a client, if windows installer contains the client/minion and master, I would suggest security needs would have issues with this, unless the client/minion by it self was still available at the same time

Personal I think `SOFTWARE\\Salt Project\\salt-minion` better choice to contain root_dir, as it allows salt-minion and salt-master to be configured differently and the semi-windows standard is  `SOFTWARE\\<Vendor>\\<Software Product>`

Reply:
I'm in agreement with @twangboy here - having a separate set of directories for config would be a shift from how salt works on Linux. It makes more sense to have `...\Salt Project\salt` contain the master/minion config, as we do on Linux, and then *those* config files can specify other locations. To your point - Salt Project is the vendor and salt is the product. The master and minion aren't independent products (see: this repository :upside_down_face:) they're just different entrypoints into the same software.

For the installer concerns, I don't know if we currently have it configured in the installer but it should be entirely possible to select which feature (master/minion) one wants to install.  Of course what that really boils down to is what entrypoints are added. I'm not even sure if the *current* approach removes any of the salt codebase -- AFAIK it just doesn't add specific entry point scripts. You can easily test this by launching the python that Salt is installed with and doing `import salt.master`.